archipelago: build from source

pyrox0 opened the pull request:
This might be one of the worst solutions to a problem(that only affects me)(the problem is that I didn't like how it's extracted from an appimage instead of being built from source) that I've ever made.

Noted issues: 
* ~~SNI(Super Nintendo Interface) is not distributed(but it should be, so that we can have compatibility with all the SNES games that Archipelago can randomize for)~~ this should now be working, but I don't have any SNES game randomizers to test with.
* ~~The A Link To The Past Enemizer(enemy randomizer) application is also not included.~~ This is now done
	* ~~Basic package code is there, but it doesn't build properly. This may need to be solved some other way than building from source, may have to just grab a binary from github releases and use that instead.~~ build issues solved

I would like to make the above 2 changes before this merges, for full functionality.

Note: This fully bypasses the upstream build system of archipelago as well as Nix's regular PEP-517 python build system. It's only used for some wrapping conveniences. 

Tested and working with a Celeste Archipelago

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pyrox0 committed:
python312Packages.pymem: init at 1.13.1

pyrox0 committed:
python312Packages.pyevermizer: init at 0.48.0

pyrox0 committed:
python312Packages.maseya-z3pr: init at 1.0.0.rc1

pyrox0 committed:
python312Packages.xxtea: init at 3.3.0

pyrox0 committed:
python312Packages.factorio-rcon-py: init at 2.1.3

pyrox0 committed:
python312Packages.zilliandomizer: init at 0.9.0

pyrox0 committed:
python312Packages.s2clientprotocol: 3.19.1.58600.0 -> 5.0.12.92440.3

pyrox0 committed:
python312Packages.s2clientprotocol: modernize

pyrox0 commented:
Nixfmt check is failing because the .NET deps.nix file for enemizer isn't formatted. Something to be looked at.

pyrox0 commented:
@corngood r.e our discussion on matrix, this is the PR.

pyrox0 committed:
sni: init at 0.0.98

pyrox0 committed:
enemizer: init at 7.1

pyrox0 committed:
archipelago: build from source

pyrox0 commented:
pinging @MinerSebas, you tested my last archipelago PR, and was wondering if you'd like to test this one as well. Feel free not to, but more testers would be good on this as it is a pretty major change.

MinerSebas commented:
Initially after ran archipelago i got this error: 
```
$./result/bin/archipelago        
Traceback (most recent call last):
  File "/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/bin/.archipelago-wrapped", line 28, in <module>
    from worlds.LauncherComponents import Component, components, Type, SuffixIdentifier, icon_paths
  File "/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/worlds/__init__.py", line 15, in <module>
    user_folder = user_path("worlds") if user_path() != local_path() else user_path("custom_worlds")
                                         ^^^^^^^^^^^
  File "/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/Utils.py", line 183, in user_path
    shutil.copytree(local_path(dn), user_path(dn), dirs_exist_ok=True, copy_function=copy_fn)
  File "/nix/store/xscbbcx260v1d9libcrc472jwx210bc0-python3-3.11.10/lib/python3.11/shutil.py", line 573, in copytree
    return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/xscbbcx260v1d9libcrc472jwx210bc0-python3-3.11.10/lib/python3.11/shutil.py", line 527, in _copytree
    raise Error(errors)
shutil.Error: [('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/custom/.gitignore', '/home/sebastian/Archipelago/data/sprites/custom/.gitignore', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/custom/.gitignore'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/custom/link.apsprite', '/home/sebastian/Archipelago/data/sprites/custom/link.apsprite', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/custom/link.apsprite'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Matty_LA.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Matty_LA.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Matty_LA.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Bowwow.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Bowwow.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Bowwow.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Mario.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Mario.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Mario.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Tarin.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Tarin.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Tarin.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Richard.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Richard.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Richard.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Bunny.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Bunny.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Bunny.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Luigi.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Luigi.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Luigi.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/alttpr/.gitignore', '/home/sebastian/Archipelago/data/sprites/alttpr/.gitignore', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/alttpr/.gitignore'")]
```

But after removing the `/home/sebastian/Archipelago` Folder it started succesfully.

One change i noticed is that it doesn't appear in my home menu anymore (Missing `archipelago.desktop` file?).
But this has the "upside" that it longer breaks initial Minecraft setup which is only shown in a console (y/n for Downloading Forge, accept EULA, etc..).

Also now that the Minecraft randomizer runs on `1.20.4`, i think the `archipelago-minecraft` package variant could updated to use `jdk21`.

keenanweaver commented:
Semi-related-but-not-at-all, do you have any plans for packaging the Archipelago-specific builds of programs? Things like [Archipelago Crispy Doom](https://archipelago.gg/tutorial/DOOM%20II/setup/en), [V6AP](https://archipelago.gg/tutorial/VVVVVV/setup/en), etc. I'm willing to test for all this stuff, but it almost seems like playing Archipelago in wine is more suited 'cause of all the prep work here.

pyrox0 commented:
> Initially after ran archipelago i got this error:
> 
> ```
> $./result/bin/archipelago        
> Traceback (most recent call last):
>   File "/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/bin/.archipelago-wrapped", line 28, in <module>
>     from worlds.LauncherComponents import Component, components, Type, SuffixIdentifier, icon_paths
>   File "/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/worlds/__init__.py", line 15, in <module>
>     user_folder = user_path("worlds") if user_path() != local_path() else user_path("custom_worlds")
>                                          ^^^^^^^^^^^
>   File "/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/Utils.py", line 183, in user_path
>     shutil.copytree(local_path(dn), user_path(dn), dirs_exist_ok=True, copy_function=copy_fn)
>   File "/nix/store/xscbbcx260v1d9libcrc472jwx210bc0-python3-3.11.10/lib/python3.11/shutil.py", line 573, in copytree
>     return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/nix/store/xscbbcx260v1d9libcrc472jwx210bc0-python3-3.11.10/lib/python3.11/shutil.py", line 527, in _copytree
>     raise Error(errors)
> shutil.Error: [('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/custom/.gitignore', '/home/sebastian/Archipelago/data/sprites/custom/.gitignore', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/custom/.gitignore'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/custom/link.apsprite', '/home/sebastian/Archipelago/data/sprites/custom/link.apsprite', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/custom/link.apsprite'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Matty_LA.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Matty_LA.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Matty_LA.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Bowwow.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Bowwow.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Bowwow.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Mario.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Mario.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Mario.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Tarin.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Tarin.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Tarin.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Richard.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Richard.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Richard.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Bunny.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Bunny.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Bunny.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/ladx/Luigi.bdiff', '/home/sebastian/Archipelago/data/sprites/ladx/Luigi.bdiff', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/ladx/Luigi.bdiff'"), ('/nix/store/59xvp93k7i8c3dggpm8858l5a7ryv01q-archipelago-0.5.0/lib/python/archipelago/data/sprites/alttpr/.gitignore', '/home/sebastian/Archipelago/data/sprites/alttpr/.gitignore', "[Errno 13] Permission denied: '/home/sebastian/Archipelago/data/sprites/alttpr/.gitignore'")]
> ```
> 
> But after removing the `/home/sebastian/Archipelago` Folder it started succesfully.
> 
> One change i noticed is that it doesn't appear in my home menu anymore (Missing `archipelago.desktop` file?). But this has the "upside" that it longer breaks initial Minecraft setup which is only shown in a console (y/n for Downloading Forge, accept EULA, etc..).
> 
> Also now that the Minecraft randomizer runs on `1.20.4`, i think the `archipelago-minecraft` package variant could updated to use `jdk21`.

interesting. I ran into that error because the shutil.copytree function(used to copy files into that player-specific directory) copies file permissions of the files it copies. and since they come from the nix store they are all marked read-only. That was one of the problems I wanted to solve with this(and did end up solving but it can't handle an existing read-only directory). I'll look into re-adding the .desktop file, should be able to be copied out or generated. I'll update the JDK as well.

As for the other programs, I'd be glad to help write packages for them, but since I don't use or play them, I wouldn't want to be a maintainer for them.

pyrox0 committed:
archipelago-minecraft: Update to jdk 21

Since this randomizer now uses Minecraft 1.20.4, this is the best JDK
version available.

getchoo commented on pkgs/development/python-modules/zilliandomizer/default.nix:
```diff
@@ -0,0 +1,40 @@
+
+  build-system = [ setuptools-scm ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    pytest-cov-stub
+  ];
```
```suggestion
  nativeCheckInputs = [ pytestCheckHook ] ++ lib.flatten (lib.attrValues optional-dependencies);
```

getchoo commented on pkgs/by-name/en/enemizer/package.nix:
```diff
@@ -0,0 +1,39 @@
+  };
+
+  patches = [
+    ./fix-build.patch
```
Would be nice to have a comment explaining why this is needed

pyrox0 commented:
Closing this as I'm not willing to keep up the development effort to ensure this works constantly. Thanks for everyone's help with this PR.
